### PR TITLE
Inventory: Send dirty lists where appropriate

### PR DESCRIPTION
**Pull request by SmallJoker**
### Goal
 * Only send modified lists to the player
 * Skip `UpdateCrafting` when nothing was changed

### Implementation
 * Bumps protocol to 38 for backwards compatibility
 * Unittests to confirm it's working as it should
 * Error which should occur to indicate erroneous sent lists
 * Possibility for incrementally sent `ItemStack`s in the future (books, much metadata)

With 200-ish more lines it would (in theory) also be possible to adapt this per-player for server-wide sent inventories, such as node- or detached inventories.

## To do

This PR is Ready for Review.

## How to test

1) Run `minetest --run-unittests` -> inventory stuff must pass
2) Play in-game. Add items, move items. Craft stuff.
3) Extended version of above: Revert one of the recent commits to re-add logging

### CODE REVIEW

File: src/client/game.cpp [hunk 1]
```diff
@@ -837,7 +836,6 @@ class Game {
 	Camera *camera = nullptr;
 	Clouds *clouds = nullptr;	                  // Free using ->Drop()
 	Sky *sky = nullptr;                         // Free using ->Drop()
-	Inventory *local_inventory = nullptr;
```
Comment (SmallJoker) [inline src/client/game.cpp]: Are you (kind reviewer) OK with this change?
 * No out-of-sync inventories
 * No memory duplication

On the cost of getting the inventory through `LocalPlayer` each time.

**Comment by SmallJoker** [comment]
So here are some new testing results. `OUT` = server, `IN` = client. Size in bytes.

**Test conditions:**
 * unified_inventory
 * 1 medium bag equipped
 * `main` filled with the first 32 items from the item sidebar

<details>
<summary><< Incremental mode</summary>
```
OUT: size=1744, time_us=45
<join msg>
OUT: size=158, time_us=6
OUT: size=158, time_us=6
OUT: size=158, time_us=7
OUT: size=158, time_us=6
OUT: size=158, time_us=7
 IN: size=1744, time_us=301
 IN: size=158, time_us=27
 IN: size=158, time_us=27
 IN: size=158, time_us=25
 IN: size=158, time_us=26
 IN: size=158, time_us=26
OUT: size=1166, time_us=40
 IN: size=1166, time_us=285
OUT: size=1166, time_us=20
 IN: size=1166, time_us=216
```
</details>

<details>
<summary><< Non-incremental (master HEAD) mode</summary>
```
OUT: size=1744, time_us=41
<join msg>
OUT: size=1744, time_us=40
OUT: size=1744, time_us=38
OUT: size=1744, time_us=36
OUT: size=1744, time_us=38
OUT: size=1744, time_us=36
 IN: size=1744, time_us=308
 IN: size=1744, time_us=244
 IN: size=1744, time_us=239
 IN: size=1744, time_us=409
 IN: size=1744, time_us=253
 IN: size=1744, time_us=239
OUT: size=1744, time_us=31
 IN: size=1744, time_us=361
OUT: size=1744, time_us=40
 IN: size=1744, time_us=380
```
</details>

<details>
<summary><< Incremental + zlib compression (not added, benefit too low)</summary>
```
OUT: size=433, time_us=128 (compressed size)
<join msg>
OUT: size=95, time_us=53
OUT: size=95, time_us=26
OUT: size=95, time_us=25
OUT: size=95, time_us=25
OUT: size=95, time_us=26
 IN: size=1744, time_us=330 (decompressed size)
 IN: size=158, time_us=35
 IN: size=158, time_us=33
 IN: size=158, time_us=32
 IN: size=158, time_us=32
 IN: size=158, time_us=31
OUT: size=357, time_us=166
 IN: size=1166, time_us=287
OUT: size=350, time_us=133
 IN: size=1166, time_us=290
```
</details>

**Summary:** (percentages compared to sending the entire inventory)
 * Sent data decreases to 10% if `main` is unaltered
 * Spent server time decreases to ~20%
 * Spent client time decreases to ~15%

However, these numbers may change quickly as soon there are books or other items with much metadata inside the inventory.

File: src/client/game.cpp (hunk 1 above)
Comment (ClobberXD) [reply src/client/game.cpp]: Getting the inventory from `LocalPlayer` seems to just be a matter of doing `player->inventory`. That's perfectly fine IMO, unless there's a disadvantage that I don't notice. Even if there was a semi-negligible disadvantage, they're heavily outweighed by the advantage of no out-of-sync inventories.

We could make the pointer private, and add a const getter though, but that's for a different discussion. :)

File: src/client/client.h [hunk 2]
```diff
@@ -277,9 +277,7 @@ class Client : public con::PeerHandler, public InventoryManager, public IGameDef
 
 	// Returns true if the inventory of the local player has been
 	// updated from the server. If it is true, it is set to false.
-	bool getLocalInventoryUpdated();
-	// Copies the inventory of the local player to parameter
-	void getLocalInventory(Inventory &dst);
+	bool updateWieldedItem();
```
Comment (ClobberXD) [inline src/client/client.h]: The comment above might need some updating

File: src/script/lua_api/l_object.cpp [hunk 3]
```diff
@@ -353,7 +353,7 @@ int ObjectRef::l_set_wielded_item(lua_State *L)
 	ItemStack item = read_item(L, 2, getServer(L)->idef());
 	bool success = co->setWieldedItem(item);
 	if (success && co->getType() == ACTIVEOBJECT_TYPE_PLAYER) {
-		getServer(L)->SendInventory(((PlayerSAO*)co));
+		getServer(L)->SendInventory((PlayerSAO*)co, true);
```
Comment (ClobberXD) [inline src/script/lua_api/l_object.cpp]:
```suggestion
		getServer(L)->SendInventory((PlayerSAO *)co, true);
```

File: src/database/database-files.cpp [hunk 4]
```diff
@@ -58,7 +58,7 @@ void PlayerDatabaseFiles::serialize(std::ostringstream &os, RemotePlayer *player
 
 	os << "PlayerArgsEnd\n";
 
-	player->inventory.serialize(os);
+	player->inventory.serialize(os, false);
```
Comment (sfan5) [inline src/database/database-files.cpp]: param should default to false so this is not needed here or anywhere else

File: src/inventory.cpp [hunk 5]
```diff
@@ -867,15 +881,32 @@ void Inventory::deSerialize(std::istream &is)
...
+			// Incrementally sent list
+			std::string listname;
+			std::getline(iss, listname, ' ');
+
+			InventoryList *list = getList(listname);
+			if (list)
+				new_lists.push_back(list);
+			else
+				errorstream << "Inventory::deSerialize(): Tried to keep list '" <<
+					listname << "' which is non-existent." << std::endl;
```
Comment (sfan5) [inline src/inventory.cpp]: braces

File: src/network/networkprotocol.h [hunk 6]
```diff
@@ -194,9 +194,11 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 		New network float format
 		ContentFeatures version 13
 		Add full Euler rotations instead of just yaw
+	PROTOCOL VERSION 38:
+		Incremental inventory sending mode
```
Comment (sfan5) [inline src/network/networkprotocol.h]: maybe mention that unknown fields(?) in inv serialization now no longer throw an error

File: src/script/lua_api/l_object.cpp (hunk 3 above)
Comment (sfan5) [reply src/script/lua_api/l_object.cpp]: if you're changing this anyway it should be a `dynamic_cast<>`
(do we use those anywhere else?)

File: src/server.cpp [hunk 7]
```diff
@@ -2595,7 +2601,8 @@ void Server::sendDetachedInventory(const std::string &name, session_t peer_id)
 
 		// Serialization & NetworkPacket isn't a love story
 		std::ostringstream os(std::ios_base::binary);
-		inv_it->second->serialize(os);
+		inv_it->second->serialize(os, false);
+		inv_it->second->setModified(false);
 
 		std::string os_str = os.str();
```
Comment (sfan5) [inline src/server.cpp]: change this to constref too, while you're at it

File: src/server.cpp [hunk 8]
```diff
@@ -2823,6 +2830,9 @@ void Server::UpdateCrafting(RemotePlayer *player)
 	if (!clist || clist->getSize() == 0)
 		return;
 
+	if (!clist->checkModified())
+		return;
```
Comment (sfan5) [inline src/server.cpp]: somehow i feel a verbosestream message would be appropriate here

File: src/inventory.cpp (hunk 5 above)
Comment (SmallJoker) [reply src/inventory.cpp]: The style rules depend on statements, not lines.

> Do not unnecessarily use braces where a single statement will do.
>
>	if (condition)
>		action();

https://www.kernel.org/doc/Documentation/process/coding-style.rst
Do you insist on adding them anyway?

**Comment by SmallJoker** [comment]
Comments solved, retested.

File: src/inventory.cpp (hunk 5 above)
Comment (sfan5) [reply src/inventory.cpp]: I don't think I've seen **multi-line** if/else statements without braces anywhere else, so it'd still be nice if you could add them.

File: src/client/client.cpp [hunk 9]
```diff
@@ -1344,25 +1344,19 @@ void Client::setPlayerControl(PlayerControl &control)
...
 	m_playeritem = item;
-	m_inventory_updated = true;
+	m_update_wielded_item = true;
 	sendPlayerItem(item);
 }
 
 // Returns true if the inventory of the local player has been
 // updated from the server. If it is true, it is set to false.
-bool Client::getLocalInventoryUpdated()
+bool Client::updateWieldedItem()
```
Comment (red-001) [inline src/client/client.cpp]: There seems to be a bunch of outdated comments?

Comment (SmallJoker) [reply src/client/client.cpp]: No, it's still valid.

File: src/server.cpp [hunk 10]
```diff
@@ -1533,21 +1533,27 @@ void Server::SendNodeDef(session_t peer_id,
...
+	NetworkPacket pkt(TOCLIENT_INVENTORY, 0, sao->getPeerID());
 
-	std::ostringstream os;
-	playerSAO->getInventory()->serialize(os);
-
-	std::string s = os.str();
+	std::ostringstream os(std::ios::binary);
+	sao->getInventory()->serialize(os, incremental);
+	sao->getInventory()->setModified(false);
+	player->setModified(true);
```
Comment (sfan5) [inline src/server.cpp]: why is the player marked as modified here?

Comment (ClobberXD) [reply src/server.cpp]: I was wondering the same thing.

Comment (SmallJoker) [reply src/server.cpp]: Inventories can be per-node or per-player. In this case we need to set the player dirty to save the inventory.

Comment (ClobberXD) [reply src/server.cpp]: Oh ok, thanks for clarifying

Comment (sfan5) [reply src/server.cpp]: But shouldn't the player be marked as modified when the inventory changes, not when it's sent?
Or does that happen at the same time anyway?

Comment (ClobberXD) [reply src/server.cpp]: `Server::handleCommand_InventoryAction` calls `Server::setInventoryModified`, which calls `Server::SendInventory`, which in turn calls `Player::setModified`. They do seem to happen at the same time.

https://github.com/minetest/minetest/blob/a0527851c95505c003c99a35fcc57c61c1d374d4/src/network/serverpackethandler.cpp#L558-L736

https://github.com/minetest/minetest/blob/a0527851c95505c003c99a35fcc57c61c1d374d4/src/server.cpp#L1243-L1263

Comment (SmallJoker) [reply src/server.cpp]: Added in both places to ensure the player is marked as modified.

**Comment by SmallJoker** [comment]
Rebased.

**Pull request merged**